### PR TITLE
Add support for enum switch kcontrols

### PR DESCRIPTION
**Pull request by ranj063**

### CODE REVIEW

File: sound/soc/soc-topology.c [hunk 1]
```diff
@@ -1364,6 +1364,10 @@ static struct snd_kcontrol_new *soc_tplg_dapm_widget_denum_create(
 
 	for (i = 0; i < num_kcontrols; i++) {
 		ec = (struct snd_soc_tplg_enum_control *)tplg->pos;
+
+		tplg->pos += (sizeof(struct snd_soc_tplg_enum_control) +
+				ec->priv.size);
+
```
Comment (plbossart) [inline sound/soc/soc-topology.c]: is this the right position?
In soc_tplg_dbytes_create, the pointer is modified after strlen+ kzalloc is checked. same in soc_tplg_dmixer_create.
But not in soc_tplg_denum_create. It's done between kzalloc and strlen check in soc_tplg_dapm_widget_dmixer_create(). 

it'd be good to have a more regular pattern across objects/widgets.

File: sound/soc/sof/topology.c [hunk 2]
```diff
@@ -1277,9 +1320,6 @@ static int sof_widget_load_pga(struct snd_soc_component *scomp, int index,
 	/* get volume control */
 	scontrol = sm->dobj.private;
 
-	/* set cmd for pga kcontrol */
-	scontrol->cmd = SOF_CTRL_CMD_VOLUME;
-
```
Comment (plbossart) [inline sound/soc/sof/topology.c:1285]: why is this removed?
and is this is correct, why is there still a volume allocated?

File: sound/soc/soc-topology.c (hunk 1 above)
Comment (ranj063) [reply sound/soc/soc-topology.c]: @plbossart ok let me make it consistent across all controls

File: sound/soc/sof/topology.c (hunk 2 above)
Comment (ranj063) [reply sound/soc/sof/topology.c:1285]: @plbossart this is removed because I moved the command setting part to load_control(). Since both switch/volume controls are the same type of control with a difference in their max values, it makes sense to do it in load_control().

I havent added support for more than 1 kcontrol for pga comp so it still can have only 1 kcontrol that should be a volume kcontrol. So we will not allocate a volume table for switch controls.

File: sound/soc/soc-topology.c (hunk 1 above)
Comment (ranj063) [reply sound/soc/soc-topology.c]: @plbossart fixed now

Comment (plbossart) [reply sound/soc/soc-topology.c]: I don't see how this is aligned now, sorry, the location of the increment is different?

**Comment by ranj063** [comment]
@plbossart is this version better?

File: sound/soc/soc-topology.c [hunk 3]
```diff
@@ -1364,6 +1365,7 @@ static struct snd_kcontrol_new *soc_tplg_dapm_widget_denum_create(
 
 	for (i = 0; i < num_kcontrols; i++) {
 		ec = (struct snd_soc_tplg_enum_control *)tplg->pos;
+
```
Comment (plbossart) [inline sound/soc/soc-topology.c]: spurious line

**Comment by ranj063** [comment]
@plbossart fixed now

**Comment by ranj063** [comment]
> looks good. @ranj063 can you send the first patch upstream? Or I can do it, as you wish

@plbossart sorry to dump it on you but could you please post it. I'm not likley going to be checking my emails for the next 10 days and so wont be able to address comments if any

**Comment by plbossart** [comment]
> @plbossart <https://github.com/plbossart> sorry to dump it on you but 
> could you please post it. I'm not likley going to be checking my emails 
> for the next 10 days and so wont be able to address comments if any

no worries :-)
